### PR TITLE
Typescript migration 17284

### PR DESCRIPTION
**Pull request by riatzukiza**
#### Summary

Migrate widgets/inputs to typescript

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/17284

#### Related Pull Requests

- Has server changes (please link here)
- Has redux changes (please link here)
- Has mobile changes (please link here)

#### Release Note

```release-note
```

### CODE REVIEW

**Comment by devinbinnie** [comment]
Hey @riatzukiza, do you have questions or need any assistance finishing this up?

**Comment by riatzukiza** [comment]
@devinbinnie Been working on this in my extra time, I made some progress, but my workstation had a failiure, I will be migrating the work I had on its hard drive to a laptop tomarrow and continue then. I was getting caught up on some bigger type errors last friday, but I think I've worked passed those and am starting to knock out all the smaller ones. I will provide more of an update tomarrow. I've been working on this thursdays and fridays.

**Comment by riatzukiza** [comment]
Hello, been a minute, there are some issues with these files that is making conversion to typescript challenging. There are several callback style methods defined on the class that seem to have totally different signatures in their definition than their callers require in their component properties. My day is comeing to an end here, I will document at least a few of these odd type mismatches before I finish for the weekend.

File: components/widgets/inputs/users_emails_input.tsx [hunk 1]
```diff
@@ -243,34 +319,32 @@ export default class UsersEmailsInput extends React.PureComponent {
...
     }
 
-    showAddEmail = (input, values, options) => {
+    showAddEmail = (input: string, _:any, options: any[]): boolean => {
         return this.props.emailInvitationsEnabled && options.length === 0 && isEmail(input);
     }
 
     onFocus = () => {
-        this.selectRef.current.handleInputChange(this.props.inputValue, {action: 'custom'});
+        this.selectRef.current?.handleInputChange({value: this.props.inputValue}, {action: 'custom'});
```
Comment (devinbinnie) [inline components/widgets/inputs/users_emails_input.tsx:339]: How come we're changing the object being passed in here?

File: components/widgets/inputs/users_emails_input.tsx [hunk 2]
```diff
@@ -300,8 +375,18 @@ export default class UsersEmailsInput extends React.PureComponent {
...
-                    loadingMessage={this.loadingMessage}
-                    onInputChange={this.handleInputChange}
+                    loadingMessage={
+                        // I don't think I can retain the same behavior here without refactoring
+                        // this function or some how changing the accepted types of `loadingMessage
+                        // property of `AsyncSelect` 
+                        this.loadingMessage
+                    }
+                    onInputChange={
+                        // I have fiddled with the type signature for this function a lot, the way it
```
Comment (devinbinnie) [inline components/widgets/inputs/users_emails_input.tsx:385]: Can you post the error you're getting?

File: components/widgets/inputs/users_emails_input.tsx [hunk 3]
```diff
@@ -25,7 +29,79 @@ import {isGuest} from 'utils/utils';
 
 import './users_emails_input.scss';
 
-export default class UsersEmailsInput extends React.PureComponent {
+import {UserProfile} from 'mattermost-redux/types/users';
+
+const AsyncSelect = require('react-select/lib/Async').default as React.ElementType<AsyncSelectProps<AppSelectOption>>; // eslint-disable-line global-require
```
Comment (devinbinnie) [inline components/widgets/inputs/users_emails_input.tsx:34]: How come we are changing the import here? I don't think this is how we should be importing things.

**Review by devinbinnie — COMMENTED** [review]
Thanks for your work @riatzukiza, I've added some comments that will hopefully help.

File: components/widgets/inputs/users_emails_input.tsx (hunk 1 above)
Comment (riatzukiza) [reply components/widgets/inputs/users_emails_input.tsx:339]: Up stream types were  not agreeing with the way these were structured. This happened several times, theses functions are basically saying that could take any of 2 or 3 types and it was making the type signatures really loopy

**Comment by riatzukiza** [comment]
> Thanks for your work @riatzukiza, I've added some comments that will hopefully help.

Much appreciated, I'll get deeper into these after I'm off work today

File: components/widgets/inputs/users_emails_input.tsx (hunk 1 above)
Comment (devinbinnie) [reply components/widgets/inputs/users_emails_input.tsx:339]: I'm not sure I understand, so you're saying that the types were different than the ones we were passing in? Can you post the error you would get if you left it alone?

File: components/widgets/inputs/users_emails_input.tsx (hunk 2 above)
Comment (riatzukiza) [reply components/widgets/inputs/users_emails_input.tsx:385]:
```
Type '() => JSX.Element' is not assignable to type '(obj: { inputValue: string; }) => string | null'.
  Type 'Element' is not assignable to type 'string'.ts(2322)
Select.d.ts(144, 3): The expected type comes from property 'loadingMessage' which is declared here on type 'IntrinsicAttributes & Props<AppSelectOption> & AsyncProps<AppSelectOption> & { children?: ReactNode; }'
```

Though I think the issue is that I was going off of what vs code was telling me was an error, I don't get any of the errors I am seeing in code when I run `make test`. The errors taht I do get seem to be tests some where else in the code, not sure whats causing those.

File: components/widgets/inputs/users_emails_input.tsx [hunk 4]
```diff
@@ -300,8 +375,18 @@ export default class UsersEmailsInput extends React.PureComponent {
                     defaultOptions={false}
                     defaultMenuIsOpen={false}
                     openMenuOnClick={false}
-                    loadingMessage={this.loadingMessage}
-                    onInputChange={this.handleInputChange}
+                    loadingMessage={
+                        // I don't think I can retain the same behavior here without refactoring
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply components/widgets/inputs/users_emails_input.tsx:379]:
```
Type '() => JSX.Element' is not assignable to type '(obj: { inputValue: string; }) => string | null'.
  Type 'Element' is not assignable to type 'string'.ts(2322)
Select.d.ts(144, 3): The expected type comes from property 'loadingMessage' which is declared here on type 'IntrinsicAttributes & Props<AppSelectOption> & AsyncProps<AppSelectOption> & { children?: ReactNode; }'
```

Though I think the issue is that I was going off of what vs code was telling me was an error, I don't get any of the errors I am seeing in code when I run `make test`. The errors taht I do get seem to be tests some where else in the code, not sure whats causing those.

File: components/widgets/inputs/users_emails_input.tsx (hunk 2 above)
Comment (riatzukiza) [reply components/widgets/inputs/users_emails_input.tsx:385]:
```
Type '(inputValue: IOptionType, action: { action: string; }) => void' is not assignable to type '(newValue: string, actionMeta: InputActionMeta) => void'.
  Types of parameters 'inputValue' and 'newValue' are incompatible.
    Type 'string' is not assignable to type 'IOptionType'.ts(2322)
Select.d.ts(173, 3): The expected type comes from property 'onInputChange' which is declared here on type 'IntrinsicAttributes & Props<AppSelectOption> & AsyncProps<AppSelectOption> & { children?: ReactNode; }'
```

**Comment by riatzukiza** [comment]
Could I get some eyes on this? The tests are all passing, dealt with some of the small issues.

File: components/widgets/inputs/users_emails_input.tsx [hunk 5]
```diff
@@ -179,8 +257,11 @@ export default class UsersEmailsInput extends React.PureComponent {
         </React.Fragment>
     );
 
-    NoOptionsMessage = (props) => {
-        const inputValue = props.selectProps.inputValue;
+    NoOptionsMessage = ({selectProps, props}: {
+        selectProps: {inputValue?: string};
+        props?: any;
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply components/widgets/inputs/users_emails_input.tsx:237]: I can't find that type

**Comment by devinbinnie** [comment]
@riatzukiza There are some lint and type-check errors as well as some conflicts with the master branch. Could you resolve those issues and then ping me? I can have a look then, thanks :)

**Comment by devinbinnie** [comment]
Closing as this is the stale and the work has been completed.

**Pull request closed without merging**